test/unittest/unittest_common_properties.cpp

```diff
@@ -101,92 +101,49 @@ TEST(NameProperty, mustStartWithAlphaNumeric_01_n) {
+    EXPECT_EQ("a(0)", nntrainer::to_string(actual));
   }
 
   {
-    InputSpec expected(
-      ConnectionSpec({Name("A"), Name("B"), Name("C")}, "addition"));
-
-    InputSpec actual;
-    nntrainer::from_string("A+ B +C", actual);
-    EXPECT_EQ("a+b+c", nntrainer::to_string(actual));
-
-    EXPECT_EQ(actual, expected);
-  }
-
-  {
-    InputSpec expected(ConnectionSpec({Name("a")}, ConnectionSpec::NoneType));
-
-    InputSpec actual;
-    nntrainer::from_string("a", actual);
-    EXPECT_EQ("a", nntrainer::to_string(actual));
+    InputConnection expected(Connection("a", 2));
 
+    InputConnection actual;
+    nntrainer::from_string("a(2)", actual);
     EXPECT_EQ(actual, expected);
+    EXPECT_EQ("a(2)", nntrainer::to_string(actual));
   }
 }
 
-TEST(InputSpecProperty, emptyString_n_01) {
+TEST(InputConnection, emptyString_n_01) {
   using namespace nntrainer::props;
-  InputSpec actual;
+  InputConnection actual;
   EXPECT_THROW(nntrainer::from_string("", actual), std::invalid_argument);
 }
 
-TEST(InputSpecProperty, combinedOperator_n_01) {
-  using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A,B+C", actual), std::invalid_argument);
-}
-
-TEST(InputSpecProperty, combinedOperator_n_02) {
-  using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A+B,C", actual), std::invalid_argument);
-}
-
-TEST(InputSpecProperty, noOperator_n_01) {
-  using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A B", actual), std::invalid_argument);
-}
-
-TEST(InputSpecProperty, noOperator_n_02) {
-  using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A B", actual), std::invalid_argument);
-}
-
-TEST(InputSpecProperty, leadingOperator_n_01) {
-  using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string(",A,B", actual), std::invalid_argument);
-}
-
-TEST(InputSpecProperty, leadingOperator_n_02) {
+TEST(InputConnection, onlyIndex_n_01) {
   using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("+A+B", actual), std::invalid_argument);
+  InputConnection actual;
+  EXPECT_THROW(nntrainer::from_string("[0]", actual), std::invalid_argument);
 }
 
-TEST(InputSpecProperty, trailingOperator_n_01) {
+TEST(InputConnection, invalidFormat_n_01) {
   using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A,B,,", actual), std::invalid_argument);
+  InputConnection actual;
+  EXPECT_THROW(nntrainer::from_string("a[0", actual), std::invalid_argument);
 }
 
-TEST(InputSpecProperty, trailingOperator_n_02) {
+TEST(InputConnection, invalidFormat_n_02) {
   using namespace nntrainer::props;
-  InputSpec actual;
-  EXPECT_THROW(nntrainer::from_string("A+B++", actual), std::invalid_argument);
+  InputConnection actual;
+  EXPECT_THROW(nntrainer::from_string("[0", actual), std::invalid_argument);
```

Comment:
consider a couple more unittests:
`a((0))`
`a((0)`
`a(0))`
`a(0)(1)`

Reply:
I think this won't pass, but this has to be checked. as the place where parsing happens has changed, I'll complement this as a new PR.

Reply:
cool.